src/common/AndroidUtils.ts

```diff
@@ -527,6 +572,206 @@ export class AndroidUtils {
+     * @param waitForBoot Optional boolean indicating whether it should wait for the device to boot up. Defaults to true.
+     */
+    public static async rebootEmulator(
+        portNumber: number,
+        waitForBoot: boolean = true
+    ): Promise<void> {
+        return AndroidUtils.executeAdbCommand('shell reboot', portNumber).then(
+            () => {
+                if (waitForBoot) {
+                    return AndroidUtils.waitUntilDeviceIsReady(portNumber);
+                } else {
+                    return Promise.resolve();
+                }
+            }
+        );
+    }
+
+    /**
+     * Determines if an emulator was launched with -writable-system parameter by looking at its emu-launch-params.txt file.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @returns A boolean indicating whether the emulator on the given port was launched with -writable-system parameter
+     */
+    public static async isEmulatorSystemWritable(
+        portNumber: number
+    ): Promise<boolean> {
+        try {
+            const emuPath = (
+                await AndroidUtils.executeAdbCommand('emu avd path', portNumber)
+            )
+                .split('\n')[0]
+                .trim();
+
+            const paramFile = path.join(emuPath, 'emu-launch-params.txt');
+
+            const launchParams = fs.readFileSync(paramFile, 'utf8');
+            return Promise.resolve(launchParams.includes('-writable-system'));
+        } catch (error) {
+            this.logger.warn(
+                `Unable to determine if emulator is system writable: ${error}`
+            );
+        }
+
+        return Promise.resolve(false);
+    }
+
+    /**
+     * Given an emulator that is running on a port, it returns the name of the AVD that is running on that port.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @returns The name of the AVD that is running on that specified port.
+     */
+    public static async fetchEmulatorNameFromPort(
+        portNumber: number
+    ): Promise<string> {
+        return AndroidUtils.executeAdbCommand(
+            'emu avd name',
+            portNumber
+        ).then((result) => result.split('\n')[0].trim());
+    }
+
+    /**
+     * Remounts adb as root with writable system access for the AVD that is running on the specified port. If the AVD currently
+     * is not launched with writable system access, this function will restart it with write access first then remounts as root.
+     *
+     * @param portNumber The ADB port of the Android virtual device.
+     * @param numRetries Optional parameter that indicates the number of retries if an error occurs. Defaults to 0 (no retries).
+     * @param retryDelay Optional parameter that indicates the milliseconds of delay before retrying again. Defaults to AndroidConfig.AdbRetryAttemptDelay.
+     */
+    public static async remountAsRootWritableSystem(
```

Comment:
I'm a little confused by the purpose of this method. It feels like we have half of the "boot a writeable emulator" code in `AndroidUtils.startEmulator()`, and the other half of it in here. Without seeing the consumption, I'm having a hard time ascertaining why it doesn't all just go in some form in `AndroidUtils.startEmulator()`.

Reply:
I guess I'm not confused by the method itself, but rather more specifically why we're putting some of the "make sure it's a writeable partition" logic in here.

Reply:
I understand exactly what you're referring to and was struggling myself too, to choose an approach over the other. At the end of the day the reasons I choose this approach were:

1. I initially had `remountAsRootWritableSystem()` take parameters for defining retry/delay values and didn't want those to be added to `startEmulator()` (this is not the case anymore since as per your feedback I have removed the parameters and move them to `executeAdbCommandWithRetry()`)

2. Putting the logic of this method into `startEmulator()` would make that method too large/complex.

3. I wasn't sure if there ever is a case for launching an emulator with `-writable-system` but not needing to also run `adb root && adb remount`. In my view the methods in `AndroidUtils` should be generic enough to serve many needs and not just tailored for specific use cases. To me, merging `remountAsRootWritableSystem` and `startEmulator()` together seemed like reducing the generality of `AndroidUtils` to better serve a specific use case. Whereas leaving them separate would keep things generic enough and would allow the consumer to decide whether they want to also `adb root && adb remount` on top of `-writable-system` (specially since for API >= 29 this would also involve disabling ABV and Verify). Again this is b/c I wasn't sure if there ever is a case for launching an emulator with `-writable-system` but not needing to also run `adb root && adb remount`. If we feel that (in addition to our current use cases) most if not all of the times, `-writable-system` will be used in conjunction with `adb root && adb remount` then I can go ahead and merge the two.

Reply:
I think we can leave it for now. I too don't want any given method to get super unwieldy or carry too many responsibilities. Let's keep it as it is.